concordium-consensus/src/Concordium/GlobalState/CooldownQueue.hs

```diff
@@ -0,0 +1,217 @@
+    Cooldowns
+        { inCooldown = Map.insertWith (+) expiry preAmt inCooldown,
+          preCooldown = Absent,
+          ..
+        }
+
+-- | Transfer the pre-pre-cooldown to pre-cooldown.
+-- If there is already an amount in pre-cooldown, the two amounts are combined.
+processPrePreCooldown :: Cooldowns -> Cooldowns
+processPrePreCooldown c@Cooldowns{prePreCooldown = Absent} = c
+processPrePreCooldown Cooldowns{preCooldown = Absent, ..} =
+    Cooldowns
+        { preCooldown = prePreCooldown,
+          prePreCooldown = Absent,
+          ..
+        }
+processPrePreCooldown Cooldowns{preCooldown = Present preAmt, prePreCooldown = Present prePreAmt, ..} =
+    Cooldowns
+        { preCooldown = Present (preAmt + prePreAmt),
+          prePreCooldown = Absent,
+          ..
+        }
+
+-- | Get the timestamp of the first cooldown that will expire, if any.
+-- (This ignores pre-cooldown and pre-pre-cooldown.)
+firstCooldownTimestamp :: Cooldowns -> Maybe Timestamp
+firstCooldownTimestamp Cooldowns{..} = fst <$> Map.lookupMin inCooldown
+
+-- | Parameters that are used to calculate the timestamps at which stake in pre-cooldown and
+--  pre-pre-cooldown is expected to be released from cooldown.
+data CooldownCalculationParameters = CooldownCalculationParameters
+    { -- | The duration of an epoch.
+      ccpEpochDuration :: Duration,
+      -- | The current epoch number.
+      ccpCurrentEpoch :: Epoch,
+      -- | The time of the next epoch transition (i.e. trigger block time).
+      ccpTriggerTime :: Timestamp,
+      -- | The epoch number of the next payday. (Must be after the current epoch.)
+      ccpNextPayday :: Epoch,
+      -- | The length of a reward period in epochs.
+      ccpRewardPeriodLength :: RewardPeriodLength,
+      -- | The current duration to for cooldowns.
+      ccpCooldownDuration :: DurationSeconds
+    }
+
+-- | Calculate the timestamp at which stake in pre-cooldown is expected to be released from
+--  cooldown. This is computed by adding the cooldown duration to the time of the next payday,
+--  where the time of the next payday is the time of the next epoch transition (i.e trigger
+--  block time) plus the duration of an epoch for each epoch between the current epoch and the
+--  payday.
+preCooldownTimestamp :: CooldownCalculationParameters -> Timestamp
+preCooldownTimestamp CooldownCalculationParameters{..} =
+    ccpTriggerTime
+        `addDuration` (fromIntegral (ccpNextPayday - ccpCurrentEpoch - 1) * ccpEpochDuration)
+        `addDurationSeconds` ccpCooldownDuration
+
+-- | Calculate the timestamp at which stake in pre-pre-cooldown is expected to be released from
+--  cooldown. If the next epoch is the next payday, this is the time of the payday after that.
+--  Otherwise, this is the same as the 'preCooldownTimestamp'.
+prePreCooldownTimestamp :: CooldownCalculationParameters -> Timestamp
+prePreCooldownTimestamp ccp@CooldownCalculationParameters{..}
+    | ccpNextPayday - ccpCurrentEpoch == 1 =
```

Comment:
Pre-pre-cooldown transitions to pre-cooldown at the snapshot epoch (1 epoch before payday) and from pre-cooldown to cooldown at the payday epoch. In this case, the snapshot epoch has already passed, so it won't enter cooldown at the payday epoch (which is the upcoming epoch), but at the payday after that.

Reply:
It seems to me, if `ccpNextPayday - ccpCurrentEpoch >= 1`, the timestamp would not be the `preCooldownTimestamp`, but the duration from the current epoch to the next payday, plus another reward period duration, plus trigger time. What am I getting wrong?

Reply:
I assume you mean `ccpNextPayday - ccpCurrentEpoch > 1`. In this case, it will enter pre-cooldown 1 epoch before the next payday (which is a future epoch) and then enter cooldown at the payday (and be in cooldown for `ccpCooldownDuration`). So the correct timestamp is trigger time (i.e. time of the start of the next epoch) plus duration from next epoch till the payday plus the cooldown duration. Which is exactly the same as `preCooldownTimestamp`.

Reply:
I'm puzzled that the `preCooldownTimestamp` and the `prePreCooldownTimestamp` are identical, except in the case that the `ccpCurrentEpoch` is one epoch before `ccpNextPayday`. I would have thought, that the `prePreCooldownTimestamp` should in most cases be further out than the `preCooldownTimestamp`.